common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,206 @@
+        RestartRX();
+      } else if (IsOnlyMarkersReceived()) {
+        // We have received just this marker
+        // In this case, we were lucky to get to LOST state in the interframe
+        // silence, assume this is the start of the frame
+        state_ = State::RECEIVING_FRAME;
+      } else {
+        // Should never end up here
+        // DMA is not working?
+        // TODO alert, safe reset
+        RestartRX();
+      }
+      break;
+    case State::WAIT_FOR_START_MARKER:
+      if (IsDataReceived()) {
+        // some junk received while waiting for the start marker,
+        // but should have been just silence
+        state_ = State::LOST;
+        RestartRX();
+      } else if (IsOnlyMarkersReceived()) {
+        // we have received the marker we are waiting for
+        state_ = State::RECEIVING_FRAME;
+      } else {
+        // Should never end up here
+        // DMA is not working?
+        // TODO alert, safe reset
+        state_ = State::LOST;
+        RestartRX();
+      }
+      break;
+    case State::RECEIVING_FRAME:
+      // end marker received, check if we got something
+      if (IsDataReceived()) {
+        // yes, we got data, thus we got the frame we can pass further
+        processReceivedData();
+        state_ = State::WAIT_FOR_START_MARKER;
+        RestartRX();
+      } else if (IsOnlyMarkersReceived()) {
+        // repeated marker char received
+        // this means we have received a 0 lenght frame
+        // ignore it and continue receiving frame bytes
+        // rx_buffer_.RestartRX(this);
+      } else {
+        state_ = State::LOST;
+        RestartRX();
+        // Should never end up here
+        // DMA is not working?
+        // TODO alert, safe reset
+      }
+      break;
+    }
+  }
+
+  // Callback method called when underlying Rx system experiences an error
+  void OnRxError([[maybe_unused]] RxError e) override {
+    switch (state_) {
+    case State::LOST:
+    case State::WAIT_FOR_START_MARKER:
+      // no change
+      break;
+    case State::RECEIVING_FRAME:
+      state_ = State::LOST;
+      RestartRX();
+      break;
+    }
+  }
+
+  // Returns the last successfully detected frame, resets the frame_available
+  // flag. I.e. if frame_available was returning true before this call, it will
+  // return false after it
+  uint8_t *TakeFrame() {
+    frame_available_ = false;
+    return frame_buf_;
+  }
+
+  // Returns the length of the last successfully detected frame
+  uint32_t get_frame_length() { return frame_buf_length_; }
+
+  // Returns true if a new frame was detected and this frame is available for
+  // read
+  bool is_frame_available() { return frame_available_; }
+
+#ifdef TEST_MODE
+  State get_state() { return state_; }
+#endif
+
+private:
+  RxBuffer &rx_buffer_;
+  State state_ = State::LOST;
+  bool frame_available_ = false;
+  uint8_t frame_buf_[FRAME_BUF_LEN];
+  uint32_t frame_buf_length_ = 0;
+  uint32_t marker_count_ = 0;
```

Comment:
What are the possible values of this? From mentally tracing the code I'm having trouble understanding whether it ever grows above 1 or 2. If it can take only a couple of values, then it's better to replace it with an enum or something, to emphasize that the state space of the current class is limited.

Reply:
Hm, I have put a bit of thinking into this today. In normal flow of things it does not go over 2, so yes, your suggestion makes total sense. 
But then I will have to deal with a case that "will never happen" if it's increasing over 2 instead of `marker_count_++`. Now in the current implementation we don't care if it becomes >2 as ProcessData will take care of that correctly. 

Do you think, should we go for it in this light? Do you have a suggestion on what how to implement it in elegant way? Override `++` operator?